plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/FileSystemExchangeSource.java

```diff
@@ -138,12 +185,125 @@ public void close()
     {
         // Make sure we will only close once
         synchronized (this) {
-            if (closed) {
+            if (!closed.compareAndSet(false, true)) {
                 return;
             }
-            closed = true;
+            files.clear();
+            Closer closer = Closer.create();
+            for (ExchangeStorageReader reader : readers.getAndSet(ImmutableList.of())) {
+                closer.register(reader);
+            }
+            try {
+                closer.close();
+            }
+            catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
         }
+    }
 
-        readers.forEach(ExchangeStorageReader::close);
+    private void createReadersIfNecessary()
+    {
+        int numberOfActiveReaders = getNumberOfActiveReaders();
+        if (numberOfActiveReaders == exchangeSourceConcurrentReaders) {
+            return;
+        }
+        if (numberOfActiveReaders > 0 && files.isEmpty()) {
```

Comment:
IntelliJ that `files.isEmpty()` is accessed in non-synchronized section. Do we really need `@GuardedBy("this")` on files?

Reply:
Also why there is `numberOfActiveReaders > 0` in this check?